### ISSUE TITLE
Ship `RabbitMQ.Client.OpenTelemetry` `1.0.0`

### ISSUE THREAD

**Issue by lukebakken** [rabbitmq/rabbitmq-dotnet-client#1728]
Thank you for using RabbitMQ.

**STOP NOW AND READ THIS** BEFORE OPENING A NEW ISSUE ON GITHUB

Unless you are CERTAIN you have found a reproducible problem in RabbitMQ or
have a **specific, actionable** suggestion for our team, you must first ask
your question or discuss your suspected issue on the mailing list:

https://groups.google.com/forum/#!forum/rabbitmq-users

Team RabbitMQ does not use GitHub issues for discussions, investigations, root
cause analysis and so on.

Please take the time to read the CONTRIBUTING.md document for instructions on
how to effectively ask a question or report a suspected issue:

https://github.com/rabbitmq/rabbitmq-server/blob/master/CONTRIBUTING.md#github-issues

Following these rules **will save time** for both you and RabbitMQ's maintainers.

Thank you.

**Comment by BOLT04** [comment]
Hello everyone,
Does anyone know the status of this issue and the whole milestone? There is only this issue on the milestone, so I'm not sure what is missing from the messaging conventions.

v7.0.0 released recently and it included some PRs for OpenTelemetry, which is awesome!

I saw recent changes related to OpenTelemetry conventions #1717 , @lukebakken is there a task list or backlog of some kind for us to know when `RabbitMQ.Client.OpenTelemetry 1.0.0` is ready to ship?

Thanks in advance 🙂!

**Comment by lukebakken** [comment]
I'm extremely busy working for customers who _pay_ Broadcom for RabbitMQ support. I am also working on higher-priority issues, such as #1750.

What would help me out is for someone, like you, to build this project from source and ensure that the OTel changes that have been made in `main` work as intended, and make sense. I don't know much about OTel so I'm relying on community members to provide input, reviews, and do testing.

**Comment by larsramberg** [comment]
Regarding this issue, is the testing done by @codedesignplus-wliscano sufficient, or are there other requirements before releasing an update OTEL package.

Happy to help out to get it accross the finish line, just not clear on what requirements there are :)

**Comment by michaelklishin** [comment]
I'd say the best way to help advance the release is to re-test everything one more time.

**Comment by lukebakken** [comment]
@michaelklishin is correct! @BOLT04 @larsramberg @StasPerekrestov if you can verify that everything is working as-is with the current `RabbitMQ.Client.OpenTelemetry`. I have neither the time nor the environment to test it myself.

**Comment by StasPerekrestov** [comment]
Could you please publish rc.2 package for testing then?

**Comment by lukebakken** [comment]
@StasPerekrestov - I just pushed `RabbitMQ.Client.OpenTelemetry.1.0.0-rc.2.nupkg`. NuGet appears to be taking its time processing the upload, but it should be there soon.

**Comment by joshdean2000** [comment]
Are there any plans to officially release this package, it has been a RC for 8 months?

**Comment by lukebakken** [comment]
> Are there any plans to officially release this package, it has been a RC for 8 months?

I have had zero reports about the rc.2 package. Once I get at least _one_ user who reports that it works as they expect, I can publish 1.0.0.

**Comment by lukebakken** [comment]
For anyone subscribed here: I have filed #1967 to review the entire OpenTelemetry tracing implementation before 7.3.0.

To be clear about what that does and does not do for this issue. It is a code audit - correctness, span lifetimes, the `IsAllDataRequested`/`HasListeners` guard pattern, context propagation edge cases, and compliance with the current messaging semantic conventions - plus fixes for anything it turns up, with regression tests. It is **not** the user report this issue has been waiting on. That still needs someone running the package in a real environment against a real backend and confirming it behaves as they expect.

What the audit should do is make such a report much cheaper to act on. Right now a single "works for me" has to carry all the weight, because the implementation has never been reviewed end to end. If the code has been audited and any defects fixed, one solid report becomes enough to ship rather than the only evidence available.

Two things worth knowing if you are testing `rc.2` today:

* `rc.2` is a snapshot from 2025-11-14. The only tracing changes to `main` since then are a copyright-year bump, so testing `rc.2` is currently a fair proxy for what would ship as 1.0.0.
* #1734 (still unmerged) adds a third `ActivitySource`, `RabbitMQ.Client.Connection`, emitting spans for connection and TCP-connection attempts. It is additive and the existing wildcard `AddSource("RabbitMQ.Client.*")` picks it up, so it will not break consumers - but it does mean 1.0.0 would emit spans that `rc.2` does not, if it lands first. Reviewing that PR found two real defects, including failed TCP connection attempts being reported to backends as successful spans.

No ship date implied by any of this. Just making the state of things visible, since this issue has been quiet on specifics for a while.

**Comment by lukebakken** [comment]
## The `OpenTelemetry.Api` version floor

#1969 (NuGet dependency review for 7.3.0) defers the `OpenTelemetry.Api` floor decision to this issue, since `RabbitMQ.Client.OpenTelemetry` is the only shipping library here whose runtime dependency is a third-party package rather than a Microsoft framework package. This issue had not actually captured that decision anywhere, so the deferral pointed nowhere. Writing it down here.

Short version: **the floor is not actually a free choice, and 1.15.3 is already correct.** What *is* an open decision is a transitive conflict it creates. Details below.

### Why the floor is forced

The rule for the three shipping libraries is that a `PackageVersion` is a *floor* imposed on every downstream consumer, so it should be the oldest supported version without known vulnerabilities.

For `OpenTelemetry.Api` the vulnerability half of that rule fully determines the answer:

* GHSA-g94r-2vxg-569j (moderate) - "Excessive memory allocation when parsing OpenTelemetry propagation headers"
* Affects `OpenTelemetry.Api` `>= 0.5.0-beta.2, < 1.15.3`
* First patched version: **1.15.3**

I verified this by restoring each candidate version and running `dotnet list package --vulnerable --include-transitive`:

| Version | Result |
|---|---|
| 1.13.1 | vulnerable (GHSA-g94r-2vxg-569j) |
| 1.14.0 | vulnerable |
| 1.15.0 | vulnerable |
| 1.15.1 | vulnerable |
| 1.15.2 | vulnerable |
| **1.15.3** | **clean** |

So 1.15.3 is not a stale pin and not a conservative guess - it is exactly the lowest version satisfying the rule. Nothing lower is permissible, and going higher (1.17.0 is current) would raise the floor for consumers with no benefit. The existing pin stands as-is.

Worth noting that this advisory is squarely in this package's path rather than incidental: the whole of `RabbitMQ.Client.OpenTelemetry` is 67 lines, and its substantive work is `Propagators.DefaultTextMapPropagator.Extract` / `.Inject` on message headers, which is the code the advisory concerns.

For the record, the 1.15.3 pin arrived in 982689358 ("Update OpenTelemetry packges to 1.15.3 due to vulnerability issues"), i.e. it was already advisory-driven. This comment confirms that reasoning still holds rather than changing anything.

### The decision that is actually open: `System.Diagnostics.DiagnosticSource` 9.0.4 vs 10.0.0

`OpenTelemetry.Api` 1.15.3 depends on `System.Diagnostics.DiagnosticSource >= 10.0.0` on every TFM it offers (net462, net8.0, net9.0, netstandard2.0; net10.0 has no dependencies). That collides with a deliberate pin in `RabbitMQ.Client`.

`RabbitMQ.Client` pins `System.Diagnostics.DiagnosticSource` at **9.0.4**, and 9.x is intentional: the connection-tracing code calls `Activity.AddException`, which was added in .NET 9, so 9.0.4 is the floor *for that API*.

The dependency requirement moved between 1.13.1 and 1.14.0:

| `OpenTelemetry.Api` | Requires `DiagnosticSource` |
|---|---|
| 1.9.0 | 8.0.0 |
| 1.10.0 - 1.13.1 | 9.0.0 |
| **1.14.0 - 1.17.0** | **10.0.0** |

The unlucky part is that the version range which is 9.x-compatible (`<= 1.13.1`) is entirely inside the vulnerable range (`< 1.15.3`). There is no version of `OpenTelemetry.Api` that is both non-vulnerable and satisfied by `DiagnosticSource` 9.x. The two constraints cannot both be met.

Consumer-visible effect, verified against the released 7.2.1 package:

```
RabbitMQ.Client 7.2.1 alone
  -> System.Diagnostics.DiagnosticSource  (not surfaced)

RabbitMQ.Client 7.2.1 + OpenTelemetry.Api 1.15.3
  -> System.Diagnostics.DiagnosticSource  10.0.0
```

So any consumer using the OpenTelemetry integration gets `DiagnosticSource` 10.0.0 regardless of the client's 9.0.4 pin. NuGet resolves to the highest floor among all constraints, so the 9.0.4 pin does not hold for them. It still holds for consumers of `RabbitMQ.Client` on its own, which is the majority.

This is not a build or runtime break - 10.0.0 is API-compatible with the 9.x surface used, and `Activity.AddException` exists in both, so nothing is wrong today. It is a question of what the pin is meant to express:

1. **Leave both pins as they are.** The 9.0.4 pin keeps its meaning for direct `RabbitMQ.Client` consumers, and OpenTelemetry users get uplifted to 10.0.0 by their own dependency. Costs nothing, but the 9.0.4 comment slightly overstates its reach.
2. **Leave the pins but document the uplift**, so the next person reading "do NOT upgrade" in `Directory.Packages.props` knows it does not apply transitively via the OpenTelemetry package.
3. **Raise `RabbitMQ.Client` to `DiagnosticSource` 10.0.0** to match what OpenTelemetry consumers already resolve. This makes the pins consistent, but raises the floor for every consumer including those not using OpenTelemetry at all, which is exactly what the "oldest supported version" rule exists to prevent. Hard to justify.

My reading is (2): keep 9.0.4, since it is still the correct floor for the API actually used and still effective for most consumers, and add a note recording that the OpenTelemetry path resolves higher. Option 3 trades a real cost for cosmetic consistency.

### Relationship to shipping 1.0.0

None of the above blocks this issue. The floor is settled and no version change is proposed for `OpenTelemetry.Api`. This is bookkeeping so that #1969 can be signed off with the `OpenTelemetry.Api` line resolved rather than deferred to an issue that did not mention it.

Also unchanged: what this issue has been waiting on is still a report from someone running the package against a real backend, as discussed above. #1967 (implementation audit) is the other prerequisite.
